Migrate `\backend` to ESM

klocke-io opened the pull request:
**What this PR does / why we need it**:

Migrates the `\backend` part of the Dashboard project to ESM

**Which issue(s) this PR fixes**:
Fixes # https://github.com/gardener/dashboard/issues/2294

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       breaking|feature|bugfix|doc|other
- target_group:   user|operator|developer|dependency
-->
``` other developer
Migrated backend from CommonJS to ESM and added a temporary transpilation step for Jest tests, which are still in CommonJS until a later migration
```

klocke-io committed:
Drop esm transpilation

klocke-io committed:
Migrate Level 0 backend entry points to ESM (.mjs): server, lib/server, lib/app

klocke-io committed:
Migrate Level 1 backend core dependencies to ESM (.mjs): auth, api, healthz, github/webhook, update app.mjs imports

klocke-io committed:
Adapt iferstructure files

klocke-io committed:
Adapt iferstructure files

klocke-io committed:
Migrate Level 2 backend core dependencies to ESM (.mjs): hooks.mjs routes/index.mjs io/index.mjs

klocke-io committed:
Remove leftovers

klocke-io committed:
Moved all level3 files from .js to .mjs

klocke-io committed:
Migrate level3

klocke-io committed:
Move files for level 4 from .js to .mjs

klocke-io committed:
Migrate level 4 from cjs to esm

klocke-io committed:
Move files for level5 from .js to .mjs

klocke-io committed:
Migrate level5 from cjs to esm

klocke-io committed:
Move files from .js to .mjs for level 6

klocke-io committed:
Migrate level6 from cjs to esm

Signed-off-by: klocke-io <niklas.klocke@sap.com>

klocke-io committed:
Move files for level6.1 from .js to .mjs: finsihing security module

klocke-io committed:
Migrate level6.1 from cjs to esm: finsihing security module

klocke-io committed:
Move files for level6.2 from .js to .mjs: finish terminal  module

klocke-io committed:
Migrate level6.2 from cjs to esm: finish terminals module

klocke-io committed:
Move files for level6.3 from .js to .mjs: finish github module

klocke-io committed:
Migrate level6.3 from cjs to esm: finish github module

klocke-io committed:
Move files for level6.4 from .js to .mjs: finish services module

klocke-io committed:
Migrate level6.4 from cjs to esm: finish services module

klocke-io committed:
Move files for level7 from .js to .mjs: cache module

klocke-io committed:
Migrate level7 from cjs to esm: cache module

klocke-io committed:
Move files for level8 from .js to .mjs: utils, markdown and logger

klocke-io committed:
Migrate level8 from cjs to esm: markdown, logger and utils  module

klocke-io committed:
Move files for level9 from .js to .mjs: config

klocke-io committed:
Migrate level9 from cjs to esm: config module

klocke-io committed:
Fix export

klocke-io committed:
Renamed files from .mjs to .js and set project type to module for esm in package.json

klocke-io committed:
Rename test, mock and fixture files from .js to .cjs

klocke-io committed:
Add build step to commonjs
Signed-off-by: klocke-io <niklas.klocke@sap.com>

klocke-io committed:
Migrate tests to use dist as test target

Signed-off-by: klocke-io <niklas.klocke@sap.com>

klocke-io committed:
Remove the build step for packages

Signed-off-by: klocke-io <niklas.klocke@sap.com>

klocke-io committed:
Commit from GitHub Actions (dependency-diagrams.yml) [skip ci]

klocke-io committed:
feat: load package once version on startup

Signed-off-by: klocke-io <niklas.klocke@sap.com>

klocke-io committed:
fix: naming conflict and whitelist

Signed-off-by: klocke-io <niklas.klocke@sap.com>

klocke-io committed:
chore: clean up

Signed-off-by: klocke-io <niklas.klocke@sap.com>

klocke-io committed:
fix: broken reference

Signed-off-by: klocke-io <niklas.klocke@sap.com>

klocke-io committed:
chore: Remove file that was accidentally committed

Signed-off-by: klocke-io <niklas.klocke@sap.com>

klocke-io committed:
chore: update .gitignore

Signed-off-by: klocke-io <niklas.klocke@sap.com>

klocke-io committed:
chore: fix license

Signed-off-by: klocke-io <niklas.klocke@sap.com>

klocke-io committed:
Revert "feat: load package once version on startup"

This reverts commit 913d9f78c588840d6090b193018bc4a7b998bc3b.

klocke-io commented on .dependency-cruiser.mjs:
```diff
@@ -242,7 +242,7 @@ export default {
      */
-    // prefix: `vscode://file/${process.cwd()}/`,
+    prefix: `vscode://file/${process.cwd()}/`,
 
```
check config which files are highlighted

klocke-io replied:
resolved in 01ab97eb

petersutter commented on backend/lib/services/terminals/index.js:
```diff
@@ -108,16 +97,14 @@ function findImageDescription (containerImage, containerImageDescriptions) {
     apiVersion,
     spec: {
-      expirationSeconds: _.get(config, ['terminal', 'serviceAccountTokenExpiration'], 43200), // default is 12h
+      expirationSeconds: _.get(gardenerConfig, ['terminal', 'serviceAccountTokenExpiration'], 43200), // default is 12h
```
why di you named it differently?

petersutter commented on backend/lib/services/terminals/index.js:
```diff
@@ -702,7 +689,7 @@ function ensureTerminalAllowed ({ method, isAdmin, body }) {
 
   // whitelist methods for terminal sessions for everybody
-  if (_.includes(['list', 'fetch', 'config', 'remove', 'heartbeat', 'listProjectTerminalShortcuts'], method)) {
+  if (_.includes(['list', 'fetch', 'terminalConfig', 'remove', 'heartbeat', 'listProjectTerminalShortcuts'], method)) {
```
I think this was wrongfully changed

klocke-io replied:
since I did not want to break the external interface of the module. 


this function used to be anonymous beforehand with config as identifier in the export. 
```
async function config ({ user, body: { coordinate: { namespace, name, target } } }) {
  return getTerminalConfig({ user, namespace, name, target })
}
```

petersutter commented on backend/lib/services/terminals/index.js:
```diff
@@ -50,34 +38,34 @@ const TargetEnum = {
+const converter = createConverter()
 
-exports.create = function ({ user, body }) {
+function create ({ user, body }) {
```
this is inconsistent to the other places (regarding the export)

petersutter commented on backend/lib/services/terminals/resources.js:
```diff
@@ -78,7 +76,4 @@ function fromNodeResource ({ metadata, status = {} }) {
-  toTerminalResource,
-  fromNodeResource,
-}
+export { toTerminalResource, fromNodeResource }
```
```suggestion
export {
  toTerminalResource,
  fromNodeResource,
}
```

petersutter commented on backend/lib/utils/index.js:
```diff
@@ -4,32 +4,32 @@
-const EQUAL = '='
-const NOT_EQUAL = '!='
+export const constants = {
+  EXISTS: '\u2203',
+  NOT_EXISTS: '!\u2203',
```
why was this changed?

klocke-io replied:
resolved in ace83580

Good catch 🙏

probably when I changed the file to individual exports I also directly exported the const's and then I would not have been needed anymore.

petersutter commented on backend/lib/utils/index.js:
```diff
@@ -4,32 +4,32 @@
+}
 
-function decodeBase64 (value) {
+export function decodeBase64 (value) {
```
also here, inconsistent change (regarding export)

klocke-io replied:
addressed in 45c92f7c

petersutter commented on backend/lib/utils/index.js:
```diff
@@ -4,32 +4,32 @@
-const NOT_EXISTS = '!∃'
-const EQUAL = '='
-const NOT_EQUAL = '!='
+export const constants = {
```
exported object is not frozen (as it was before)

klocke-io replied:
resolved in ace83580

Good catch 🙏

probably when I changed the file to individual exports I also directly exported the const's and then I would not have been needed anymore.

petersutter commented on backend/lib/watches/leases.js:
```diff
@@ -57,6 +55,3 @@ module.exports = (io, informer, { signal }) => {
   informer.on('update', object => handleEvent({ type: 'MODIFIED', object }))
 }
-
-// exported for testing
```
comment got lost and `loadOpenIssuesAndComments` is not exported anymore under test

klocke-io replied:
resolved in f3a45630

petersutter commented on backend/package.json:
```diff
@@ -20,16 +21,13 @@
+  "exports": "./lib/server.js",
   "scripts": {
-    "serve": "cross-env NODE_ENV=development node server.js",
+    "serve": "cross-env NODE_ENV=development node --experimental-vm-modules server.js",
```
remove `--experimental-vm-modules` for the regular `serve`

petersutter commented on backend/rollup.config.js:
```diff
@@ -0,0 +1,34 @@
+    input: 'server.js',
+    cache: false,
+    output: {
+      dir: `${process.cwd()}/dist/`,
```
see my other comment regarding `process.cwd()`

klocke-io committed:
chore: remove .mjs highlighting

klocke-io committed:
fix: remove check for unused variable

klocke-io committed:
chore: change export structure

klocke-io commented on backend/__tests__/watches.spec.cjs:
```diff
@@ -409,7 +409,9 @@ describe('watches', function () {
 
     it('should should load issues and comments of all issues', async function () {
-      const { loadOpenIssuesAndComments } = watches.leases.test
+      // todo exported directly
+      // const { loadOpenIssuesAndComments } = watches.leases.test
+      const { loadOpenIssuesAndComments } = require('../dist/lib/watches/leases.cjs')
```
resolved in cc38b8c9

klocke-io commented on backend/__tests__/hooks.spec.cjs:
```diff
@@ -127,7 +127,8 @@ describe('hooks', () => {
         expect(cache.initialize.mock.calls[0][0]).toBe(informers)
 
         expect(io).toHaveBeenCalledTimes(1)
-        expect(io.mock.calls[0]).toEqual([server, cache])
```
resolved in da469198

klocke-io commented on backend/lib/watches/leases.js:
```diff
@@ -57,6 +55,3 @@ module.exports = (io, informer, { signal }) => {
   informer.on('update', object => handleEvent({ type: 'MODIFIED', object }))
 }
-
-// exported for testing
```
resolved in cc38b8c9

klocke-io commented on backend/__tests__/hooks.spec.cjs:
```diff
@@ -127,7 +127,8 @@ describe('hooks', () => {
 
         expect(io).toHaveBeenCalledTimes(1)
-        expect(io.mock.calls[0]).toEqual([server, cache])
+        // ToDo checking for dead variable
+        // expect(io.mock.calls[0]).toEqual([server, cache])
```
resolved in da469198

klocke-io committed:
fix: freeze object

klocke-io committed:
fix: naming

klocke-io committed:
fix: httpError destructuring

klocke-io committed:
chore: correct misplaced comment

klocke-io committed:
chore: fix format

klocke-io commented on backend/__tests__/hooks.spec.cjs:
```diff
@@ -8,19 +8,19 @@
+const watches = require('../dist/lib/watches')
 
-const createHooks = require('../lib/hooks')
+const hookPkg = require('../dist/lib/hooks')
```
yes this issue occurs a couple of times. 

beforehand the import from hooks was directly destructured, leaving the hooks identifier available for use with in the code itself. thats the case here as well, see l24

klocke-io committed:
chore: restore first line

klocke-io committed:
chore: improve import aliases

klocke-io committed:
chore: remove needless --experimental-vm-modules flag

klocke-io committed:
chore: restore dropped empty line in the license header

klocke-io committed:
fix: undo unintended adaptation

klocke-io committed:
chore: replace process.cwd() with import.meta.url approach

klocke-io commented on backend/lib/app.js:
```diff
@@ -30,7 +26,7 @@ for (const ctor of [Object, Function, Array, String, Number, Boolean]) {
 
 // resolve pathnames
-const PUBLIC_DIRNAME = resolve(join(__dirname, '..', 'public'))
+const PUBLIC_DIRNAME = resolve(join(process.cwd(), 'public'))
```
Totally agree, addressed in 84e558b5

klocke-io committed:
feat: configure ESLint import resolution for @gardener-dashboard packages

- Add shared eslint-import-resolver-local.cjs at root level
- Configure backend ESLint to resolve @gardener-dashboard/* imports to packages/
- Update frontend ESLint to use shared resolver
- Remove duplicate resolver files
- Enable proper import resolution for scoped packages in both frontend and backend

klocke-io committed:
chore: remove accidentally commited file

klocke-io committed:
chore: adapt export structure for clarity

klocke-io committed:
fix: eslint exclude file name from .js to .cjs

klocke-io committed:
chore: add comment for clarity

klocke-io committed:
fix: reference issue

klocke-io committed:
chore: rename for consistency

klocke-io committed:
chore: remove unnecessary assignment

klocke-io commented on backend/lib/io/helper.js:
```diff
@@ -4,187 +4,178 @@
+import { simplifyObjectMetadata } from '../utils/index.js'
+
+const { isHttpError } = httpErrors
+const kubernetesClient = kubeClient
```
Migration nonsense, my bad 😥

resolved in 5c19feca

klocke-io committed:
chore: rename for consistency

klocke-io commented on backend/lib/routes/info.js:
```diff
@@ -4,17 +4,27 @@
+import { fileURLToPath } from 'url'
+import logger from '../logger/index.js'
+import { decodeBase64 } from '../utils/index.js'
+import kubeClientModule from '@gardener-dashboard/kube-client'
```
agree, addressed in 2f79752a

klocke-io committed:
chore: add eslint-plugin-import-newlines

klocke-io committed:
chore: run yarn lint --fix after adding eslint-plugin-import-newlines

klocke-io commented on backend/lib/services/terminals/index.js:
```diff
@@ -4,40 +4,28 @@
+import gardenerConfig from '../../config/index.js'
+import { getClusterCaData } from '../shoots.js'
+import kubeClient from '@gardener-dashboard/kube-client'
+import { decodeBase64, getConfigValue, getSeedNameFromShoot } from '../../utils/index.js'
```
done 😉

see 5c8f3243 and 6340472d

klocke-io committed:
feat: import package.json instead of direct access

klocke-io committed:
chore: rename for clarity

klocke-io commented on backend/lib/routes/index.js:
```diff
@@ -4,27 +4,41 @@
 
 if (_.get(config, ['frontend', 'features', 'terminalEnabled'], false)) {
-  module.exports['/terminals'] = require('./terminals')
+  routes['/terminals'] = terminalsRoute
```
See my comment in https://github.com/gardener/dashboard/pull/2494#discussion_r2227930726

klocke-io commented on backend/lib/io/helper.js:
```diff
@@ -4,187 +4,178 @@
+    OBJECT_SIMPLE: 1,
+    OBJECT_ORIGINAL: 2,
+  }),
+  synchronizeFactory (kind, options = {}) {
```
So that the functions belong to one object which can be spied as before through the re export into helper const.

klocke-io committed:
chore: rename for clarity

klocke-io commented on backend/lib/services/projects.js:
```diff
@@ -4,20 +4,21 @@
+import cache from '../cache/index.js'
+const { dashboardClient } = kubeClient
+const { PreconditionFailed, InternalServerError } = httpErrors
+const { getProject, getProjects } = cache
```
addressed in f6e0db2c

klocke-io committed:
chore: adapt export structure for clarity

klocke-io committed:
fix: address accidental rename

klocke-io commented on backend/lib/services/shoots.js:
```diff
@@ -264,9 +264,9 @@ const patchAnnotations = async function ({ user, namespace, name, annotations })
 }
-exports.patchAnnotations = patchAnnotations
+export { patchAnnotations }
 
-exports.remove = async function ({ user, namespace, name }) {
+export async function remove ({ user, namespace, name }) {
```
<img width="860" height="254" alt="image" src="https://github.com/user-attachments/assets/8e5b91d4-3915-4d84-a605-08ee09d88f91" />
export in newline should be addressed

klocke-io committed:
chore: adapt export structure for clarity

klocke-io committed:
chore: adapt export structure for clarity

klocke-io committed:
chore: adapt export structure for clarity

klocke-io committed:
Merge branch 'master' into backendToEsm

# Conflicts:
#	.ci/verify
#	.pnp.cjs
#	.yarn/cache/rollup-npm-4.40.2-23636f79f1-cbe9b76689.zip
#	.yarn/cache/rollup-npm-4.41.1-8674a5e6b8-c4d5f22573.zip
#	.yarn/cache/rollup-npm-4.44.2-3468b15ea1-5ada4fd03e.zip
#	backend/lib/services/cloudprofiles.js
#	backend/package.json
#	diagram/backend-lowlevel-dependency-diagram.html
#	yarn.lock

klocke-io committed:
chore: add missing packages

petersutter reviewed (approved):
/lgtm

klocke-io committed:
fix: add build-test-target step to verify script
